cpp/include/cudf/aggregation.hpp

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <vector>
+
+/**
+ * @file aggregation.hpp
+ * @brief Representation for specifying desired aggregations from
+ * aggregation-based APIs, e.g., groupby, reductions, rolling, etc.
+ *
+ * @note Not all aggregation APIs support all aggregation operations. See
+ * individual function documentation to see what aggregations are supported.
+ *
+ */
+
+namespace cudf {
+namespace experimental {
+/**
+ * @brief Base class for abstract representation of an aggregation.
+ */
+class aggregation;
+
+/// Factory to create a SUM aggregation
+std::unique_ptr<aggregation> make_sum_aggregation();
+
+/// Factory to create a MIN aggregation
+std::unique_ptr<aggregation> make_min_aggregation();
+
+/// Factory to create a MAX aggregation
+std::unique_ptr<aggregation> make_max_aggregation();
+
+/// Factory to create a COUNT aggregation
+std::unique_ptr<aggregation> make_count_aggregation();
+
+/// Factory to create a MEAN aggregation
+std::unique_ptr<aggregation> make_mean_aggregation();
+
+/// Factory to create a MEDIAN aggregation
+std::unique_ptr<aggregation> make_median_aggregation();
+
+/**
+ * @brief Factory to create a QUANTILE aggregation
+ *
+ * @param quantiles The desired quantiles
+ * @param interpolation The desired interpolation
+ */
+std::unique_ptr<aggregation> make_quantile_aggregation(
```

Comment:
Should likely rename to `make_quantile_aggregation_args` or `make_quantile_agg_args` or `make_quantile_args`. Basically something that means it's the args.

Reply:
I'd argue the factory is not making the "args", it's making the aggregation itself. 

A `quantile` aggregation is inseparable from its arguments. It's meaningless to request a `quantile` without specifying _which_ quantiles you want and how to interpolate them. 

Likewise with standard deviation, we'd do:
```
unique_ptr<aggregation> make_stddev_aggregation(int ddof);
```

not 
```
unique_ptr<aggregation> make_stddev_aggregation_args(int ddof);
```

Plus, then the naming would be inconsistent with the factories that don't take arguments.